book-content/chapters/02-ide-superpowers.md

```diff
@@ -0,0 +1,579 @@
+ *
+ * @param obj - The object to log.
+ *
+ * @example
+ * ```ts
+ * logValues({ a: 1, b: 2 });
+ * // Output:
+ * // a: 1
+ * // b: 2
+ * ```
+ */
+
+const logValues = (obj: any) => {
+  for (const key in obj) {
+    console.log(`${key}: ${obj[key]}`);
+  }
+};
+````
+
+The `@param` tag is used to describe the parameters of the function. The `@example` tag is used to provide an example of how the function can be used, using markdown syntax.
+
+There are many, many more tags available for use in JSDoc comments. You can find a full list of them in the [JSDoc documentation](https://jsdoc.app/).
+
+Adding JSDoc comments is a useful way to communicate the purpose and usage of your code, whether you're working on a library, a team, or your own personal projects.
+
+### Exercises
+
+#### Exercise 1: Adding Documentation for Hovers
+
+Here's a simple function that adds two numbers together:
+
+```typescript
+const myFunction = (a: number, b: number) => {
+  return a + b;
+};
+```
+
+In order to understand what this function does, you'd have to read the code.
+
+Add some documentation to the function so that when you hover over it, you can read a description of what it does.
+
+<Exercise title="Exercise 1: Adding Documentation For Hovers" filePath="/src/016.5-ide-superpowers/042-adding-tsdoc-comments-for-hovers.problem.ts"></Exercise>
+
+#### Solution 1: Adding Documentation for Hovers
+
+In this case, we'll specify that the function "Adds two numbers together". We can also use an `@example` tag to show an example of how the function is used:
+
+```typescript
+/**
+ * Adds two numbers together.
+ * @example
+ * myFunction(1, 2);
+ * // Will return 3
+ */
+
+const myFunction = (a: number, b: number) => {
+  return a + b;
+};
+```
+
+Now whenever you hover over this function, the signature of the function along with the comment and full syntax highlighting for anything below the `@example` tag:
+
+```
+// hovering over myFunction shows:
+
+const myFunction: (a: number, b: number) => number
+
+Adds two numbers together.
+
+@example
+
+myFunction(1, 2);
+
+// Will return 3
+```
+
+While this example is trivial (we could, of course, just name the function better), this can be an extremely important tool for documenting your code.
+
+## Navigating with Go To Definition and Go To References
+
+The TypeScript server also provides the ability to navigate to the definition of a variable or declaration.
+
+In VS Code, this "Go to Definition" shortcut is used with `Command + Click` on a Mac or `F12` on Windows for the current cursor position. You can also right click and select "Go to Definition" from the context menu on either platform. For the sake of brevity, we'll use the Mac shortcut.
+
+Once you've arrived at the definition location, repeating the `Command + Click` shortcut will show you everywhere that the variable or declaration is used. This is called "Go To References". This is especially useful for navigating around large codebases.
```

Comment:
Show Call Hierarchy is pretty awesome too. Maybe worth adding a short section on it too?